Self-Hosted: provide migration guide to help move from native to kubernetes (#89)
packagist/private-packagist-docs: docs/self-hosted/kubernetes-migration-guide.md

### Backup the PostgreSQL database

At the end of this step, you should have a `packagist_db.sql` file in your working directory.

```
docker exec -it packagist-postgres pg_dump --clean --if-exists -U postgres -d packagist_db > packagist_db.sql
docker stop packagist-postgres
```

### Backup the Redis database

At the end of this step, you should have a `packagist_redis.rdb` file in your working directory.

```
docker exec -it packagist-redis /usr/local/bin/redis-backup.sh
docker cp packagist-redis:/data/dump.rdb packagist_redis.rdb
docker stop packagist-redis
```

### Start Private Packagist again (optional)

Now that you have exported all the data, you may restart the Private Packagist application. This will allow your developers
to continue with their work while you import the data in the Private Packagist Self-Hosted Kubernetes setup.

```
docker start packagist-postgres packagist-redis packagist-ui packagist-repo packagist-web packagist-worker    
```

Please note that any changes made in Private Packagist will now be lost and will have to be reapplied once everyone switches to Private Packagist Self-Hosted Kubernetes.

## Installing Private Packagist Self-Hosted Kubernetes

> If you haven't done so already, now is the time to follow the [installation guide](./kubernetes.md) to install Private Packagist Self-Hosted Kubernetes.

Author: Exactly, if they do it on the same machine. Then docker is gone. There is still a way back by uninstall kURL and rerunning the native installer but that is rather painful.

The guide already recommends to set up K8s before starting the migration `we recommend that you set up your Private Packagist Self-Hosted Kubernetes instance on a different machine before you start the migration process.`

## Restore application data in Kubernetes

All commands assume that you are logged in to the server where Private Packagist Self-Hosted Kubernetes is installed,
your working directory contains the three files generated by the previous steps, and that

> Private Packagist Self-Hosted Kubernetes is installed in the Kubernetes namespace "default".

Reviewer: We should recommend to use the default namespace or not?

Maybe not add -n default to all commands but switch the namespace globally? Dont know, just thinking out loud

Author: +1 on removing the namespace. Makes them easier to port with someone uses a different namespace. I would assume anyone not using the embedded cluster won't use the default namespace. Definitely wouldn't recommend that

> 

Reviewer: would maybe explicitly state that the files have to be copied to the new server, if two server setup...

I wonder if someone really wants to migrate on the same server?

Author: It does already say that the working directory needs to contain the three files

> To avoid any data inconsistency, we highly recommend going through the steps below in order and only restart the application

Reviewer: ```suggestion
To avoid any data inconsistency, we highly recommend going through the steps below in order and only start the application
```

once all the data has been restored.

### Stop the Private Packagist application

```
kubectl scale deployment ui repo worker --replicas 0 -n default
```

> 

Reviewer: this can take a bit. Maybe tell them how they can verify that?

### Restore the PostgreSQL database

Follow the instructions below to restore the PostgreSQL database in the cluster.
If you are using your own PostgreSQL instance, use the `packagist_db.sql` file directly to import the data into your instance.

Please note that the backup will overwrite any existing data in the database and any changes you made during setup will be lost.

```
kubectl scale statefulset postgres --replicas 0 -n default
kubectl delete pvc postgres-data-postgres-0 -n default
kubectl scale statefulset postgres --replicas 1 -n default

> cat packagist_db.sql | kubectl exec -it -n default statefulsets/postgres -- psql -U postgres -d packagist_db

Reviewer: Pretty tricky to review this now. Cause based on the conversation we had yesterday, we need to explain a lot about kubectl and/or how to verify that the commands worked?
Doesnt this need a bit time to scale up and down? in practice?

Follow these steps if you are using Private Packagist Self-Hosted with an embeded Kubernetes cluster
```
export PV_NAME=$(kubectl get pvc -n default redis-data-redis-0 -ojsonpath='{.spec.volumeName}')
kubectl scale statefulset -n default redis --replicas 0
rm /var/openebs/local/$PV_NAME/appendonly.aof
cp packagist_redis.rdb /var/openebs/local/$PV_NAME/dump.rdb
kubectl scale statefulset -n default redis --replicas 1
```

### Restore the dist and artifact files

Please note that depending on the size of the `packagist_storage.tar.gz` file and the type of blob storage that you have configured
it can take several minutes for this command to finish.

```
kubectl scale statefulset -n default ui --replicas 1
kubectl cp -n default packagist_storage.tar.gz ui:/tmp/packagist_storage.tar.gz
kubectl exec -n default ui /srv/manager/bin/console packagist:self-hosted:migrate-storage import /tmp/packagist_storage.tar.gz && rm /tmp/packagist_storage.tar.gz 
```

### Start the Private Packagist application

> Once your data has been restored. Restart the application, log in to Private Packagist, and verify that your organization
> and packages are there again.

Reviewer: ```suggestion
Once your data has been restored. Start the application:
```

```
kubectl scale deployment ui --replicas 1 -n default

> ```

Reviewer: Log in to (the new?) Private Packagist, and verify that your organization and packages are there again.

> kubectl scale deployment ui --replicas 1 -n default
> ```

Reviewer: ```suggestion
kubectl scale deployment ui --replicas 1 -n default
```

Log in to (the new?) Private Packagist, and verify that your organization and packages are there again.
```

```

### Import the dist and artifact files

Importing the dist and artifact files requires the UI pods to be back online. Start them with the command below. This will take a few seconds.
You can use the second command to verify once 2/2 pods are running.
```
kubectl scale statefulset ui --replicas 1
kubectl get pods -w
```

Please note that depending on the size of the `packagist_storage.tar.gz` file and the type of blob storage that you have configured
it can take several minutes for this command to finish.

```
export UI_POD=$(kubectl get pods --no-headers -o custom-columns=":metadata.name"|grep ui-)
kubectl cp packagist_storage.tar.gz $UI_POD:/tmp/packagist_storage.tar.gz -c ui
kubectl exec $UI_POD -c ui -- /bin/sh -c "/srv/manager/bin/console packagist:self-hosted:migrate-storage import /tmp/packagist_storage.tar.gz && rm /tmp/packagist_storage.tar.gz" 
```

### Start the Private Packagist application

Once your data has been restored. Start the application:

```
kubectl scale deployment ui --replicas 1
```

This can take a few minutes. You can run the command below to see when all pods are back up and running.

```
kubectl get pods -w
```

> Once the all pods are back up and running. Log in to Private Packagist and verify that your organization is shown in the UI.

Reviewer: ```suggestion
Once all the pods are back up and running, log in to Private Packagist and verify that your organization is shown in the UI.
```
Or can this be two sentences? Dont think so

> Ideally, run a `composer update` command in one of your projects to assert that the Composer repository run as expected.

Reviewer 2: ```suggestion
Ideally, run a `composer update` command in one of your projects to assert that the Composer repository responds as expected.
```

# Migrating to Private Packagist Self-Hosted Kubernetes
##

## Requirements

To start the migration process, your Private Packagist Self-Hosted Replicated Native instance needs to run version 1.12.5 or newer.
To reduce the disruption to a minimum, we recommend that you set up your Private Packagist Self-Hosted Kubernetes
instance on a different machine before you start the migration process.

Please note that it is possible to install Private Packagist Self-Hosted Kubernetes on the same machine as your current
Private Packagist Self-Hosted Installation. You can do so after you follow the instructions below to back up your existing data
and before you start restoring the application data. However, the Private Packagist Self-Hosted installation process will uninstall
certain components used by your current Private Packagist Self-Hosted installation. 

## Backing up the data

Private Packagist Self-Hosted uses three different storages. A PostgreSQL database used as primary storage,

> a Redis database used as secondary storage, cache, and job queue, and a file/blob storage for uploaded artifacts and dist

Author: How so? Secondary storage doesn't mean cache?

> a Redis database used as secondary storage, cache, and job queue, and a file/blob storage for uploaded artifacts and dist

Reviewer 2: Secondary to me suggests it stands behind the first. I don't really know what that means at all, backups? Would prefer that term would be replaced with something more concrete.

```
kubectl scale statefulset ui --replicas 1
kubectl get pods -w
```

Please note that depending on the size of the `packagist_storage.tar.gz` file and the type of blob storage that you have configured
it can take several minutes for this command to finish.

```
export UI_POD=$(kubectl get pods --no-headers -o custom-columns=":metadata.name"|grep ui-)
kubectl cp packagist_storage.tar.gz $UI_POD:/tmp/packagist_storage.tar.gz -c ui
kubectl exec $UI_POD -c ui -- /bin/sh -c "/srv/manager/bin/console packagist:self-hosted:migrate-storage import /tmp/packagist_storage.tar.gz && rm /tmp/packagist_storage.tar.gz" 
```

### Start the Private Packagist application

Once your data has been restored. Start the application:

```
kubectl scale deployment ui --replicas 1
```

This can take a few minutes. You can run the command below to see when all pods are back up and running.

```
kubectl get pods -w
```

Once the all pods are back up and running. Log in to Private Packagist and verify that your organization is shown in the UI.
Ideally, run a `composer update` command in one of your projects to assert that the Composer repository and any of the

> background workers run as expected.

Author: My thinking was that a successful `composer update` run would indicate that everything is ok but maybe I don't even need to mention the background workers.

> background workers run as expected.

Author: Updated

> background workers run as expected.

Reviewer 2: Thanks, that's good!
